### PR TITLE
IGNITE-20117 Implement index backfill process

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-20117

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.ignite.internal.storage.BinaryRowAndRowId;
+import org.apache.ignite.internal.storage.ReadResult;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.util.Cursor;
+
+/** {@link BinaryRowAndRowId} chooser for index building. */
+class BuildIndexRowVersionChooser {
+    private final PartitionDataStorage storage;
+
+    /** Timestamp of activation of the catalog version in which the index created. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java:34]: fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        this.startBuildingIndexActivationTs = startBuildingIndexActivationTs;
+    }
+
+    /**
+     * Collects binary versions of a row to build an index.
+     *
+     * <p>Index selection algorithm:</p>
+     * <ul>
+     *     <li>For writeCommitted with commitTs > activationTs(BUILDING), we will ignore.</li>
+     *     <li>For writeCommitted with commitTs <= activationTs(BUILDING), we will take the latest version.</li>
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java:60]: fix it

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooserTest.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+/** For {@link BuildIndexRowVersionChooser} testing. */
+public class BuildIndexRowVersionChooserTest extends IgniteAbstractTest {
+    private static final int TABLE_ID = 1;
+
+    private static final int PARTITION_ID = 0;
+
+    private static final long CREATE_INDEX_ACTIVATION_TS_MILLS = 100;
+
+    private static final long START_BUILDING_INDEX_ACTIVATION_TS_MILLS = 100;
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooserTest.java:53]: Fix it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexStatus;
+import org.apache.ignite.internal.storage.BinaryRowAndRowId;
+import org.apache.ignite.internal.storage.ReadResult;
+import org.apache.ignite.internal.storage.RowId;
+import org.apache.ignite.internal.util.Cursor;
+
+/** {@link BinaryRowAndRowId} chooser for index building. */
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java:30]: Apply it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                    if (commitTs(readResult) > startBuildingIndexActivationTs) {
+                        continue;
+                    } else if (takenLatestVersionOfWriteCommitted) {
+                        break;
+                    } else {
+                        takenLatestVersionOfWriteCommitted = true;
+                    }
+                }
+
+                result.add(new BinaryRowAndRowId(readResult.binaryRow(), rowId));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java:92]: I think it's worth doing this later.
Adds TODOs.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+     *     <li>For writeCommitted with commitTs > activationTs(BUILDING), we will ignore.</li>
+     *     <li>For writeCommitted with commitTs <= activationTs(BUILDING), we will take the latest version.</li>
+     *     <li>For writeIntent with beginTs >= activationTs(REGISTERED), we will ignore.</li>
+     *     <li>For writeIntent with beginTs < activationTs(REGISTERED), we take.</li>
+     * </ul>
+     *
+     * @param rowId Row ID of interest.
+     */
+    List<BinaryRowAndRowId> chooseForBuildIndex(RowId rowId) {
+        try (Cursor<ReadResult> rowVersionCursor = storage.scanVersions(rowId)) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/BuildIndexRowVersionChooser.java:68]: Add comment.

**Review by rpuch — APPROVED** [review]

**Pull request merged**